benchmarks/halo_finder/halo_finder.cpp

```diff
@@ -0,0 +1,192 @@
+  std::cout.flush();
+
+  std::ifstream input;
+  if (!binary)
+    input.open(filename);
+  else
+    input.open(filename, std::ifstream::binary);
+  ARBORX_ASSERT(input.good());
+
+  int num_points = 0;
+  std::vector<float> x, y, z;
+  if (!binary)
+  {
+    input >> num_points;
+
+    x.reserve(num_points);
+    y.reserve(num_points);
+    z.reserve(num_points);
+
+    auto read_float = [&input]() {
+      return *(std::istream_iterator<float>(input));
+    };
+    std::generate_n(std::back_inserter(x), num_points, read_float);
+    std::generate_n(std::back_inserter(y), num_points, read_float);
+    std::generate_n(std::back_inserter(z), num_points, read_float);
+  }
+  else
+  {
+    input.read(reinterpret_cast<char *>(&num_points), sizeof(int));
+
+    x.resize(num_points);
+    y.resize(num_points);
+    z.resize(num_points);
+    input.read(reinterpret_cast<char *>(x.data()), num_points * sizeof(float));
+    input.read(reinterpret_cast<char *>(y.data()), num_points * sizeof(float));
+    input.read(reinterpret_cast<char *>(z.data()), num_points * sizeof(float));
+  }
+  input.close();
+  std::cout << "done\nRead in " << num_points << " points" << std::endl;
+
+  std::vector<ArborX::Point> v(num_points);
+  for (int i = 0; i < num_points; i++)
+  {
+    v[i] = {x[i], y[i], z[i]};
+  }
+
+  return v;
+}
+
+template <typename... P, typename T>
+auto vec2view(std::vector<T> const &in, std::string const &label = "")
+{
+  Kokkos::View<T *, P...> out(
+      Kokkos::view_alloc(label, Kokkos::WithoutInitializing), in.size());
+  auto const start = std::chrono::high_resolution_clock::now();
+  Kokkos::deep_copy(out, Kokkos::View<T const *, Kokkos::HostSpace,
+                                      Kokkos::MemoryTraits<Kokkos::Unmanaged>>{
+                             in.data(), in.size()});
+  auto const end = std::chrono::high_resolution_clock::now();
+  std::chrono::duration<double> elapsed_seconds = end - start;
+  printf("deep copy       : %10.3f\n", elapsed_seconds.count());
+  return out;
+}
+
+int main(int argc, char *argv[])
+{
+  using ExecutionSpace = Kokkos::Cuda;
+  using MemorySpace = Kokkos::CudaSpace;
+
+  Kokkos::ScopeGuard guard(argc, argv);
+
+  std::cout << "ArborX version: " << ArborX::version() << std::endl;
+  std::cout << "ArborX hash   : " << ArborX::gitCommitHash() << std::endl;
+
+  namespace bpo = boost::program_options;
+
+  std::string filename;
+  bool binary;
+  float linking_length;
+  int min_size;
+
+  bpo::options_description desc("Allowed options");
+  // clang-format off
+    desc.add_options()
+        ( "help", "help message" )
+        ( "filename,f", bpo::value<std::string>(&filename), "filename containing data" )
+        ( "binary,b", bpo::bool_switch(&binary)->default_value(false), "binary file indicator")
+        ( "linking-length,l", bpo::value<float>(&linking_length), "linking length (radius)" )
+        ( "min-size,s", bpo::value<int>(&min_size)->default_value(2), "minimum halo size")
```

Comment:
Need to bring `verify()` back before the merge. We'll need to run `query()` the second time to produce the graph, and compare `ccs` against that.